### PR TITLE
Add TotalScoreSnapshot

### PR DESCRIPTION
**Pull request by MTsatsarov**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/818
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/819
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/817

Added two new properties to Participant entity: 
int TotalScoreSnapshot
Datetime? TotalScoreSnapshotModifiedOn

These properties will be updated/calculated everytime user submits a submission.

Two new Background jobs added.
1. Background job that will update participant total score once per day.
2. Background job that will remove Participant scores if there are more than one for specific problem , for practice or compete. It will remain only the one with highest points.

### CODE REVIEW

**Review by shsimeonova — APPROVED** [review]

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs [hunk 1]
```diff
@@ -245,9 +250,10 @@ private void SaveParticipantScore()
                     {
                         this.participantScoresData.AddBySubmissionByUsernameAndIsOfficial(
                             this.submission,
-                            participant.UserName,
+                            participant.User.UserName,
```
Comment (gogo4ds) [inline Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:253]: Why not directly participant.UserName?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs (hunk 1 above)
Comment (MTsatsarov) [reply Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:253]: The participant has no userName, before it had because there is Select which takes IsOfficial and UserName.

Comment (gogo4ds) [reply Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:253]: This will make extra query to get the User. Eager loading is the better approach here.

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs [hunk 2]
```diff
@@ -166,7 +165,16 @@ protected override void ProcessTestsExecutionResult(IExecutionResult<TestResult>
...
             this.UpdateResults();
         }
-
+        
+        private static void UpdateParticipantTotalScore(Participant participant)
+        {
+            participant.TotalScoreSnapshot =
+                participant.Scores.Where(ps => !ps.Problem.IsDeleted)
+                    .Select(ps => ps.Points)
+                    .Sum();
```
Comment (gogo4ds) [inline Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:174]: This again will make db request to retrieve the Scores.

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs [hunk 3]
```diff
@@ -220,11 +230,6 @@ private void SaveParticipantScore()
 
                 var participant = this.participantsData
                     .GetByIdQuery(this.submission.ParticipantId.Value)
-                    .Select(p => new
-                    {
-                        p.IsOfficial,
-                        p.User.UserName
-                    })
                     .FirstOrDefault();
```
Comment (gogo4ds) [inline Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:233]: Include all the necessary data with single request to avoid lazy loading
```
var participant = this.participantsData
    .GetByIdQuery(this.submission.ParticipantId.Value)
    .Select(p => new
    {
        Participant = p,
        UserName = p.User.UserName,
        TotalScore = p.Scores
            .Where(ps => !ps.Problem.IsDeleted)
            .Select(ps => ps.Points)
            .Sum()
    })
    .FirstOrDefault();
```

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs (hunk 2 above)
Comment (MTsatsarov) [reply Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs]: Done

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs (hunk 3 above)
Comment (MTsatsarov) [reply Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs]: Done

File: Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs [hunk 4]
```diff
@@ -167,6 +166,12 @@ protected override void ProcessTestsExecutionResult(IExecutionResult<TestResult>
             this.UpdateResults();
         }
 
+        private static void UpdateParticipantTotalScore(Participant participant, int totalScore)
+        {
+            participant.TotalScoreSnapshot = totalScore;
```
Comment (gogo4ds) [inline Open Judge System/LocalWorker/OJS.LocalWorker/OjsSubmissionProcessingStrategy.cs:171]: This should be in one transaction with adding/updating the participant score entity. We could have discrepancy between the snapshot and the actual score if something fails..
I suggest integrating this logic into to the logic of the `SaveParticipantScore()` method.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 5]
```diff
@@ -12,14 +11,17 @@
...
 
         public ParticipantScoresDataService(
             IEfGenericRepository<ParticipantScore> participantScores,
-            IParticipantsDataService participantsData)
+            IParticipantsDataService participantsData,
+            IEfGenericRepository<Participant> participantRepository)
         {
             this.participantScores = participantScores;
             this.participantsData = participantsData;
+            this.participantRepository = participantRepository;
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:24]: Can't you use` IParticipantsDataService participantsData` instead of the repository?

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 6]
```diff
@@ -65,8 +68,15 @@ public void ResetBySubmission(Submission submission)
...
                     p.IsOfficial,
-                    p.User.UserName
+                    p.User.UserName,
+                    TotalScore = p.Scores.Any()
+                        ? p.Scores
+                            .Where(ps => !ps.Problem.IsDeleted)
+                            .Select(ps => ps.Points)
+                            .Sum()
+                        : 0
                 })
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:80]: This is duplicated above. Maybe extract a dedicated model for this operations and put in a service, where it can be reused.

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 7]
```diff
@@ -119,31 +131,41 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
...
 
         public void UpdateBySubmissionAndPoints(
             ParticipantScore participantScore,
             int? submissionId,
-            int submissionPoints)
+            int submissionPoints,
+            Participant participant,
+            int totalScore)
         {
+            participant.TotalScoreSnapshot = totalScore + submissionPoints;
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:162]: Make sure the Snapshot is updated properly for Retests by:
1. Single submission
2. Problem
3. Contest

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 5 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:24]: The  IParticipantsDataService participantsData does not contains methods for Update() and SaveChanges()

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 7 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 6 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 5 above)
Comment (gogo4ds) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:26]: You can see how other data services make `Update` and add it.

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 6 above)
Comment (gogo4ds) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: I still see this duplicated, just with a model.

Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Фиьед

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 5 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 8]
```diff
@@ -119,31 +119,39 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
...
 
         public void UpdateBySubmissionAndPoints(
             ParticipantScore participantScore,
             int? submissionId,
-            int submissionPoints)
+            int submissionPoints,
+            Participant participant,
+            int totalScore)
         {
+            participant.TotalScoreSnapshot = (totalScore - participantScore.Points) + submissionPoints;
```
[the remark this answers is not in the document]
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 9]
```diff
@@ -119,31 +119,39 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
...
                 ParticipantId = submission.ParticipantId.Value,
                 ProblemId = submission.ProblemId.Value,
                 SubmissionId = submission.Id,
-                ParticipantName = username,
+                ParticipantName = userName,
                 Points = submission.Points,
-                IsOfficial = isOfficial
+                IsOfficial = participant.IsOfficial
             });
+            participant.TotalScoreSnapshot = totalScore + submission.Points;
```
[the remark this answers is not in the document]
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                Participant = participant,
+                IsOfficial = participant.IsOfficial,
+                UserName = participant.User.UserName,
+                TotalScore = participant.Scores.Any()
+                    ? participant.Scores
+                        .Where(ps => !ps.Problem.IsDeleted)
+                        .Select(ps => ps.Points)
+                        .Sum()
+                    : 0
```
[the remark this answers is not in the document]
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs]: Done

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            participant => new ParticipantScoreDataModel
+            {
+                Participant = participant,
+                IsOfficial = participant.IsOfficial,
+                UserName = participant.User.UserName,
+                TotalScore = participant.Scores
+                    .Where(ps => !ps.Problem.IsDeleted)
+                    .Select(ps => ps.Points)
+                    .DefaultIfEmpty(0)
+                    .Sum()
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs:28]: Do you still need this?

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs [hunk 12]
```diff
@@ -376,13 +399,19 @@ private double CalculateTimeInMinutesFromParticipationStartToLastSubmission(Part
                 .Select(ps => new
                 {
                     ParticipantScore = ps,
-                    ProblemMaxPoints = ps.Problem.MaximumPoints
+                    ProblemMaxPoints = ps.Problem.MaximumPoints,
+                    Particinapnt = ps.Participant,
+                    totalScore = ps.Participant.Scores
+                        .Where(scores => !scores.Problem.IsDeleted)
+                        .Select(scores => scores.Points)
+                        .Sum()
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:407]: And this?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 12 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:407]: Yes because there is a build error that says: Try specifying the type arguments explicitly.

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs (hunk 11 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs:28]: Yes, this is the totalScore of the participant without the current submission points. After that depending on number of things the points will be updated, but i need the value before update, in case that it needs to subtract the points for the current problem.

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 12 above)
Comment (gogo4ds) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs:407]: You cannot remove the totalScore from here?

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs (hunk 11 above)
Comment (gogo4ds) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs:28]: This too I don't see it used anywhere

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs (hunk 12 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs]: Done

File: Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs (hunk 11 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/ParticipantScores/Models/ParticipantScoreDataModel.cs]: Fixed

**Review by gogo4ds — APPROVED** [review]

**Comment by gogo4ds** [comment]
Redis cache (https://github.com/SoftUni-Internal/OpenJudgeSystem/pull/831) should be merged before this.

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs [hunk 13]
```diff
@@ -119,31 +118,37 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
...
         }
 
         public void UpdateBySubmissionAndPoints(
             ParticipantScore participantScore,
             int? submissionId,
-            int submissionPoints)
+            int submissionPoints,
+            Participant participant)
         {
+            participant.TotalScoreSnapshot = (participant.TotalScoreSnapshot - participantScore.Points) + submissionPoints;
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:146]: Is it guaranteed that `participant.TotalScoreSnapshot` will always be greater or equal to `participantScore.Points`?

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs [hunk 14]
```diff
@@ -31,42 +31,5 @@ public class ParticipantResultViewModel
...
-            participant => new ParticipantResultViewModel
-            {
-                ParticipantUsername = participant.User.UserName,
-                ParticipantFirstName = participant.User.UserSettings.FirstName,
-                ParticipantLastName = participant.User.UserSettings.LastName,
-                ProblemResults = participant.Scores
-                    .Where(sc => !sc.Problem.IsDeleted && sc.Problem.ProblemGroup.ContestId == contestId)
-                    .AsQueryable()
-                    .Select(ProblemResultPairViewModel.FromParticipantScoreAsFullResult)
-            };
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: Why did you remove those? I see that this logic is moved inside the controller. Can't you leave it here (with small refactoring of the queries if needed)

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs (hunk 14 above)
Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: image
 It was moved inside the controller because of this error.

Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: This logic was moved to the controller because of this error: The type arguments cannot be inferred from the usage. Try specifying the type arguments explicitly. Candidates are:   System.Collections.Generic.IEnumerable<TResult> Select<Participant,TResult>(this System.Collections.Generic.IEnumerable<OJS.Data.Models.Participant>, System.Func<OJS.Data.Models.Participant,int,TResult>) (in class Enumerable)   System.Collections.Generic.IEnumerable<TResult> Select<Participant,TResult>(this System.Collections.Generic.IEnumerable<OJS.Data.Models.Participant>, System.Func<OJS.Data.Models.Participant,TResult>) (in class Enumerable)

File: Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs (hunk 13 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs:146]: Yes, this logic is inside UpdateBydSubmissionAndPoints, therefore in worst case the participant.TotalScoreSnapshot will be equal to participantScore.Points. On the other hand the submissionPoints will always be greater than participantScore.points

File: Open Judge System/Web/OJS.Web/Controllers/TempController.cs [hunk 15]
```diff
@@ -374,6 +374,9 @@ public async Task<ActionResult> MigrateContestCategoryFromRemoteJudge(string id)
 
         public ActionResult UpdateParticipantsTotalScore()
         {
+            // This request causes 504 (Timeout error)
+            return this.Content("UPDATED");
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/Controllers/TempController.cs:378]: "UPDATED" is not correct. "Unsupported" seems more legit. Also add comment in the code that the query will work if executed in the database directly.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs [hunk 16]
```diff
@@ -498,41 +511,82 @@ public ActionResult StatsChart(int contestId)
...
+            }
+
             var participants = this.participantsData
                 .GetAllByContestAndIsOfficial(contest.Id, official)
-                .AsNoTracking();
+                .OrderByDescending(p => p.TotalScoreSnapshot)
+                .ThenBy(p => p.TotalScoreSnapshotModifiedOn)
+                .Skip(participantsToSkip)
+                .Take(participantsToTake)
+                .ToList();
```
Comment (gogo4ds) [inline Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:535]: Try not to do `.ToList` here, as this will get all the properties from db. Also you will make extra calls from lazy loading of User and Scores, when you construct the `ParticipantResultViewModel` later on in the code.
You should `.ToList` after the projection to the resulting model.
This will probably fix the exception you mentioned that made you move logic inside the controller.

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs (hunk 14 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: You can fix this 💯 
Try `.AsQueryable`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs (hunk 16 above)
Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:535]: If we remove .ToList() the returned data is not ordered. This is likely because when we remove .ToList() the query remains in "deferred state", next manipulations of the query is likely to interfere with the ordering

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs (hunk 14 above)
Comment (MTsatsarov) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: image
 It is not working, it shows the same error

File: Open Judge System/Data/OJS.Data/IOjsDbContext.cs [hunk 17]
```diff
@@ -58,12 +59,18 @@ public interface IOjsDbContext : IDisposable
 
         int SaveChanges();
 
+        Task<int> SaveChangesAsync();
+
         void ClearDatabase();
 
+        void DbExecuteSqlCommand(string query);
+
+        void ExecuteSqlCommandWithTimeout(string query, int timeoutInSeconds);
```
Comment (gogo4ds) [inline Open Judge System/Data/OJS.Data/IOjsDbContext.cs:68]: Revert these changes. Since the temp method is unsupported, those will not be used.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs (hunk 14 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Contests/ViewModels/Results/ParticipantResultViewModel.cs:58]: `.AsQueryable` should be called before the `.Select`

File: Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs (hunk 16 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:535]: This seems strange. I would like to check it out with you, when you have the chance. Maybe we can figure out what is happening, or/and learn something new.

Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs:517]: Fixed problem with N+1 queries and getting all the properties.

File: Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs [hunk 18]
```diff
@@ -139,6 +144,36 @@ public void Delete(IEnumerable<Participant> participantsForDeletion)
...
+                   SELECT SUM(ps.Points)
+               FROM ParticipantScores ps
+                   JOIN Problems pr ON ps.ProblemId = pr.Id AND pr.IsDeleted = 0
+               WHERE ps.ParticipantId = p.Id
+                   ), 0)
+               FROM OrderedParticipants op
+                   JOIN Participants p ON op.Id = p.Id;
+
+               SET @Offset = @Offset + @BatchSize;
+               END;";
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs:172]: It is good idea to extract variables for `ps.Points`, `ps.ProblemId` and the rest with `nameof()`. 
This way we will have compile time notion in the code that these properties are used in this query. If changed, the person working on it will see it, as now this can be easily missed and broken.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs (hunk 18 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 19]
```diff
@@ -146,6 +152,26 @@ public void UpdateTotalScoreSnapshotOfParticipants()
             this.participantsData.UpdateTotalScoreSnapshot();
         }
 
+        public void RemoveParticipantMultipleScores()
+        {
+            var participants = 
+                this.scoresDataService.GetAll()
+                .GroupBy(ps => new { ps.IsOfficial, ps.ProblemId, ps.ParticipantId })
+                .Where(ps=>ps.Count() > 1)
+                .ToList();
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs:161]: This is `participantScores` not `participants`

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 20]
```diff
@@ -146,6 +152,26 @@ public void UpdateTotalScoreSnapshotOfParticipants()
...
+            
+            using (var scope = TransactionsHelper.CreateTransactionScope())
+            {
+                foreach (var participant in participants)
+                {
+                    var participantsToRemove = participant.OrderByDescending(ps => ps.Points).Skip(1).ToList();
+                    this.scoresDataService.Delete(participantsToRemove);
+
+                }
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs:170]: Why this is in Transaction scope?

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 21]
```diff
@@ -146,6 +152,26 @@ public void UpdateTotalScoreSnapshotOfParticipants()
...
+                .GroupBy(ps => new { ps.IsOfficial, ps.ProblemId, ps.ParticipantId })
+                .Where(ps=>ps.Count() > 1)
+                .ToList();
+            
+            using (var scope = TransactionsHelper.CreateTransactionScope())
+            {
+                foreach (var participant in participants)
+                {
+                    var participantsToRemove = participant.OrderByDescending(ps => ps.Points).Skip(1).ToList();
+                    this.scoresDataService.Delete(participantsToRemove);
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs:168]: This will make separate db request for each participant. Can we make it more efficient, by deleting all at once or in batches (but provided they are not expected to be many, I say at once)

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs (hunk 19 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs:161]: Correct, we have multiple participantScores not multiple participants.

File: Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs [hunk 22]
```diff
@@ -146,27 +146,31 @@ public void Delete(IEnumerable<Participant> participantsForDeletion)
...
+                   SELECT SUM(ps.{points})
                FROM ParticipantScores ps
-                   JOIN Problems pr ON ps.ProblemId = pr.Id AND pr.IsDeleted = 0
-               WHERE ps.ParticipantId = p.Id
+                   JOIN Problems pr ON ps.{problemId} = pr.{nameof(ParticipantScore.Id)} AND pr.{nameof(Problem.IsDeleted)} = 0
+               WHERE ps.{nameof(ParticipantScore.ParticipantId)} = p.{nameof(Participant.Id)}
                    ), 0)
                FROM OrderedParticipants op
-                   JOIN Participants p ON op.Id = p.Id;
+                   JOIN Participants p ON op.{nameof(Participant.Id)} = p.{nameof(Participant.Id)};
```
Comment (gogo4ds) [inline Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs:173]: nameof(Participant.Id) is used many times, why not extract it as well?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs (hunk 20 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs:170]: Because delete is calling SaveChanges and i think is better idea to make only one transaction for all participantScores instead of many.

Comment (gogo4ds) [reply Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs]: You can achieve this without transaction scope

File: Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs (hunk 22 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Data/Participants/ParticipantsDataService.cs]: Done

File: Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs (hunk 20 above)
Comment (MTsatsarov) [reply Open Judge System/Services/OJS.Services/Business/Participants/ParticipantsBusinessService.cs]: Done
